WIP - Templates

Huddie opened the pull request:
Looking for preliminary feedback. I'm thinking of moving regex matching to an extension. I originally had it that the IssueTemplateHelper returned the templates to the caller and let it deal with it how it wished. I just then felt that since every view that wants to open a new issue needs to go through the same process that maybe its better to pass the view controller in and keep the code all in one place. Since it starts with an extension on the client the client isn't being passed around. 

Anyways let me know what to fix.

TODO:
- [ ] Add Spinner 
- [ ] Remove template details before passing it through (Text between the dashes ( --- ) at the beginning of a template file)

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+import Squawk
+
+struct IssueTemplate {
+    let title:String
```
Nit: spacing after `:`

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+enum IssueTemplateHelper {
+
+    static private func matchesForRegexInText(
+        regex: String!,
```
What about `matches(forRegex:inText:)`?
And why are string types force unwrapped?

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        text: String!) -> [String] {
+
+        do {
+            let regex = try NSRegularExpression(pattern: regex, options: [])
```
FYI you can omit the `options` parameter as that has a default expression of `[]`.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+
+            let results = regex.matches(
+                in: text,
+                options: [],
```
Here too

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+                range: NSMakeRange(0, nsString.length))
+            return results.map { nsString.substring(with: $0.range)}
+
+        } catch let error as NSError {
```
The compiler likes this? Cool.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    static func getNameAndDescriptionFromTemplateFile(file: String) -> (name: String?, about: String?) {
```
`getNameAndDescription(fromTemplateFile:)`?

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+
+        let names = IssueTemplateHelper.matchesForRegexInText(regex: "(?<=name:).*", text: file)
+        let abouts = IssueTemplateHelper.matchesForRegexInText(regex: "(?<=about:).*", text: file)
+        let name = names.count > 0
```
What about `names.first?.trimmingCharacters(in: .whitespaces)`? Then you skip some checks.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        return (name, about)
+    }
+
+    static func showIssueAlert(
```
It would be cool if a function creates the alert and returns it, so it can be tested.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        let alertView = UIAlertController(
+            title: NSLocalizedString("New Issue", comment: ""),
+            message: NSLocalizedString("Choose Template", comment: ""),
+            preferredStyle: .alert
```
Not sure if we want to have this be an alert instead of an action sheet?

Huddie replied:
Let me know. I can switch

BasThomas replied:
@rnystrom?

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+                            template: template.template,
+                            signature: template.title == "Bug Report" ? .bugReport : .sentWithGitHawk
+                            ) else {
+                                Squawk.showGenericError()
```
Can we show a custom error with Squawk? Maybe add an error message / code to debug?

BasThomas replied:
Or at least an `assert` is this is not expected to happen.

Huddie replied:
Ya I'll switch to assert

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+                                return
+                        }
+                        viewController.delegate = delegate
+                        let navController = UINavigationController(rootViewController: viewController)
```
Not sure if we want to pull out this presenter logic. cc @rnystrom

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        alertView.addAction(
+            UIAlertAction(
+                title: "Dismiss",
+                style: UIAlertActionStyle.cancel,
```
You can omit `UIAlertActionStyle`; it is inferred.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+
+        alertView.addAction(
+            UIAlertAction(
+                title: "Dismiss",
```
This should be localized.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+                title: "Dismiss",
+                style: UIAlertActionStyle.cancel,
+                handler: { _ in
+                    alertView.dismiss(animated: true, completion: nil)
```
You can omit the `completion` parameter; it defaults to `nil`.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+                    alertView.dismiss(animated: true, completion: nil)
+            })
+        )
+        mainViewController.present(alertView, animated: true, completion: nil)
```
Same here.

BasThomas replied:
And make them `trueUnlessReduceMotionEnabled` maybe? :)

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        self.fetchFile(
+            owner: owner,
+            repo: repo,
+            branch: "master",
```
Is this expected to be hardcoded?

BasThomas replied:
I don't think so. I think you can even not have a `master` branch; can we get the default branch somehow? cc @rnystrom

Huddie replied:
@BasThomas "If you create templates on other branches, they will not be available for collaborators to use."
[Link](https://help.github.com/articles/about-issue-and-pull-request-templates/)

Huddie replied:
Actually, if someone knows how to get default branch easily that would be good. I know we have access to it in the branch switcher, not sure if its a quick get or not

Huddie replied:
Figured this out, got it from the RepositoryDetails

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    private func createTemplateWith(
```
Neat!

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    func createNewIssue(
```
Maybe split some of this up to make it more readable; it does a lot, making it hard to parse.

BasThomas commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,220 @@
+                    // Sort lexicographically
+                    let sortedTemplates = templates.sorted(by: {$0.title < $1.title })
+
+                    if sortedTemplates.count > 0 {
```
This logic can be pulled out, for instance.

Huddie replied:
Pulled it out, is it better?

BasThomas commented on Classes/New Issue/NewIssueTableViewController.swift:
```diff
@@ -51,6 +51,7 @@ final class NewIssueTableViewController: UITableViewController, UITextFieldDeleg
     @IBOutlet var titleField: UITextField!
     @IBOutlet var bodyField: UITextView!
 
+    private var template: String = ""
```
Nit: you can omit the `: String`; it is inferred.

Sherlouk replied:
Would we not just rather this be nil if there is no template? Feels like a tidier solution in my opinion

Huddie replied:
@Sherlouk Updated!

Sherlouk replied:
@Huddie Swish! If you give us a shout once you're looking for another review and I can do a full sweep at that point, hard to keep track of all these individual discussions at this point 😅

BasThomas commented on Classes/New Issue/NewIssueTableViewController.swift:
```diff
@@ -85,6 +86,26 @@ final class NewIssueTableViewController: UITableViewController, UITextFieldDeleg
+        owner: String,
+        repo: String,
+        template: String,
+        signature: IssueSignatureType? = nil) -> NewIssueTableViewController? {
```
Nit: newline for the return type?

BasThomas commented on Classes/New Issue/NewIssueTableViewController.swift:
```diff
@@ -85,6 +86,26 @@ final class NewIssueTableViewController: UITableViewController, UITextFieldDeleg
+
+        let storyboard = UIStoryboard(name: "NewIssue", bundle: nil)
+
+        let viewController = storyboard.instantiateInitialViewController() as? NewIssueTableViewController
```
Should we exit scope if we can't do this? Can't imagine that that would be the case unless somebody removes the storyboard or changes its name 😛

BasThomas replied:
Would be awesome! Can also do this in a follow up sometime.

BasThomas commented on Classes/Repository/RepositoryViewController.swift:
```diff
@@ -142,21 +142,16 @@ ContextMenuDelegate {
-        else {
-            Squawk.showGenericError()
-            return nil
+        let action = UIAlertAction(title: "New Issue", style: UIAlertActionStyle.default) { _ in
```
`"New Issue"` should be localized.

Huddie replied:
I changed this in both createWithTemplate(..) and create(..), I'm going to move all the localized strings to the Constants file as well

BasThomas commented on Classes/Repository/RepositoryViewController.swift:
```diff
@@ -142,21 +142,16 @@ ContextMenuDelegate {
-        else {
-            Squawk.showGenericError()
-            return nil
+        let action = UIAlertAction(title: "New Issue", style: UIAlertActionStyle.default) { _ in
```
And you can omit the `UIAlertActionStyle` again

Sherlouk commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,233 @@
+                            owner: details.owner,
+                            repo: details.repo,
+                            template: template.template,
+                            signature: template.title == "Bug Report" ? .bugReport : .sentWithGitHawk
```
How does everyone feel about this check?

Is this essentially saying that if *anybody* creates a template in their repository called "Bug Report" then it will use the bug report template? Or is this limited just to the GitHawk repo?

I ask this because the Bug Report signature is pretty specific to GitHawk even going as far to include TestFlight information which isn't really applicable if it's a bug report on _pluck random repo out of the air_ equally the device info probably isn't that important.

Might be missing something here though!

Huddie replied:
Ya that a good point. I meant to revisit this. Maybe check if repo =  GitHawk or owner = GitHawkApp

BasThomas replied:
Yep, that seems fair. Maybe create a `GitHawkApp` repo constant somewhere and check that. Didn't think of that.

Huddie replied:
Ya I'll create that. 👍

Sherlouk replied:
Could go fully overpowered and add it as an option to the markdown table at the top of template 😉but nah think the simple and easy solution is as above!

Huddie replied:
Haha @Sherlouk if we added more details into the bug report that could be a cool way to allows other repos to make use of the general bug report info. But i don't know what useful data we could add to the bug report that benefits any repo.

Huddie replied:
Switched checking titles to checking repo owner, now only GitHawkApp will have the bug report dump.

Sherlouk commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,233 @@
+            owner: owner,
+            repo: repo,
+            branch: "master",
+            path: ".github/ISSUE_TEMPLATE/\(filename)") { (result) in
```
Have we double checked if GitHub is specific about the caps on this? I know previously when I investigated this I had way too many failures causes by the fact GitHub was case insensitive when checking for this!

(This could be different now though!)

Sherlouk replied:
Also think @BasThomas picked up on the hardcoded branch here, maybe needs to be the default branch?

Super nit: should we move these constants to somewhere together? (The two path constants)

Huddie replied:
Good point. I remember saying it must be under ISSUE_TEMPLATE but ill recheck.

BasThomas replied:
It is case insensitive; we should check that.

Huddie replied:
@BasThomas as in you can have .github/issue_template ? Or the current code is case insensitive?

BasThomas replied:
As in you can have `.github/issue_template`

Huddie replied:
@BasThomas @Sherlouk I did a bare bones test adding a file under .github/ISSUE_TEMPLATE and then under .github/issue_template. First one worked, the second one did not. I think it may have to be under ISSUE_TEMPLATE unless someone knows of a counter example

Sherlouk replied:
So for context my logic was coming from [this official article](https://help.github.com/articles/about-issue-and-pull-request-templates/) but is more about the singular templates as apposed to the new pick a template chooser (as I mentioned in the accompanying issue, this is step 2).

> Issue template filenames are not case sensitive, and can have an extension such as .md or .txt.

If you've tested it we might be safe for this iteration, worst case we raise a bug and fix it at a later date! Ship and iterate!

Sherlouk replied:
Think this get's a bit confusing because it depends on exactly what the repo is doing but I think as per my comment in the accompanying issue ("step 2") is that single issue templates are not case sensitive and have a lot of different formats but potentially the new multi issue templates require the all caps directory?

https://help.github.com/articles/manually-creating-a-single-issue-template-for-your-repository/

Good article above on the topic showing that the markdown files can be in a bunch of places, and I believe are definitely not case sensitive

Huddie replied:
@Sherlouk Ah I think I understand. The actual filename isn't an issue here because we look at all files in the folder and request those filenames, so they will always match perfectly. The only issue would be the ISSUE_TEMPLATE folder as that's the only hardcoded path. I agree that maybe we ship and fix if there's a bug report. Waiting on Ryan for a few things before I remove the WIP and finish up

Huddie replied:
@Sherlouk Just saw the link, so there's a difference between single issues vs multiple issues it appear (As you mentioned above). I still agree that maybe we go with this and as we learn more about what works etc we fix. If we can't find your templates we still show a blank New Issue screen so it won't have negative effects.

Sherlouk commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,233 @@
+            owner: owner,
+            repo: repo,
+            branch: "master",
+            path: ".github/ISSUE_TEMPLATE") { (result) in
```
Same comment as previous, just would like to test case sensitivity on this

Sherlouk commented:
Think this is an ace addition to the app, making it super easy to make issues can only help productivity! Good work!

Huddie commented:
> Think this is an ace addition to the app, making it super easy to make issues can only help productivity! Good work!

Thanks! I really appreciate that

BasThomas commented:
Yessss super cool!

Sherlouk commented on Classes/Settings/SettingsViewController.swift:
```diff
@@ -162,19 +162,17 @@ NewIssueTableViewControllerDelegate, DefaultReactionDelegate {
+    }
+
+    func showTemplateOptions() {
+        client.createNewIssue(
```
To make sure I'm clued up on what's happening here, so tapping on "Report a Bug" is now going to ask you which template to use "Bug Report" or "Feature Request", is that correct?

I'm easy but is this the functionality we want? Maybe we should rename the button or maybe default to the Bug Report template, or if people are fine with this we can keep the same!

Huddie replied:
I don't have a preference,  switching to Report Bug/ Feature request sounds good or defaulting to bug. Currently, we don't have a quick link for feature request in settings. Maybe now we just offer both, bug report + feature request.

Sherlouk replied:
It's a weird one, as a user I want to be able to do both easily! I'm thinking a button rename would be the best, but I'm lost on the naming for it!

BasThomas replied:
"Create an issue"?

Huddie commented:
* Uses the repos default branch now
* Moved path into constants

Huddie commented:
Added completion block so the caller knows when the template UIAlertView is about to be displayed. This allows for showing a spin indicator in the calling view. I implemented that in the settings view.

ijm8710 commented:
Curious, so the way it works is to provide a framework on what you’re reporting? Is it the user’s job to replace the template with their report? Is there any way to set it up so it prompts you, this is what you enter here without the need to delete? This is super cool either way but was just curious so that the template is still a reference but also streamlines the input

Huddie commented:
>Curious, so the way it works is to provide a framework on what you’re reporting? Is it the user’s job to replace the template with their report? Is there any way to set it up so it prompts you, this is what you enter here without the need to delete? This is super cool either way but was just curious so that the template is still a reference but also streamlines the input

@ijm8710 if you could move this comment to #2393 this thread is more of the release of the feature, that thread is about the feature itself

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Sherlouk commented on Classes/New Issue/IssueTemplates.swift:
```diff
@@ -0,0 +1,230 @@
+            preferredStyle: .alert
+        )
+
+        for template in templates {
```
Correct me if I'm wrong, there appears to be no way to create an empty issue anymore? Is that correct?

As per the image below, GitHub has a button which allows you to just create an empty issue without a template!

![image](https://user-images.githubusercontent.com/15193942/48300953-ebd04380-e4dd-11e8-8db8-d509b636efba.png)

Huddie commented:
@Sherlouk Looking for a second review. If you could give some regex feedback I'd appreciate it. I'm pretty new to it. Maybe @BrianLitwin has some improvement suggestions on that.

Huddie commented:
@rnystrom I think Bas pinged you on some reviews above. I believe the 2 topics were:

- AlertView vs ActionSheet
   - https://github.com/GitHawkApp/GitHawk/pull/2400#discussion_r230915222
- If we want to pull out presenter logic
   - https://github.com/GitHawkApp/GitHawk/pull/2400#discussion_r230888506
Feedback would be appreciated when you get a chance as I think this is close to being ready

jdisho commented on Classes/Settings/SettingsViewController.swift:
```diff
@@ -154,19 +154,25 @@ NewIssueTableViewControllerDelegate, DefaultReactionDelegate {
+        showTemplateOptions()
+    }
+
+    func showTemplateOptions() {
```
In my understanding, you are not using this `func` anywhere else except `onReportBug`. Why not putting the implementation in `onReportBug`?

jdisho commented on Classes/Utility/String+Regex.swift:
```diff
@@ -0,0 +1,34 @@
+
+extension String {
+    func matches(
+        regex: String) -> [String] {
```
IMO, the new line makes it less unreadable.

jdisho commented on Classes/Views/SpinnerTableCell.swift:
```diff
@@ -0,0 +1,35 @@
+
+final class SpinnerTableCell: StyledTableCell {
+
+    let activity = UIActivityIndicatorView(activityIndicatorStyle: .gray)
```
Are you using this outside this class? If not, make it private :)

jdisho commented on Classes/Views/SpinnerTableCell.swift:
```diff
@@ -0,0 +1,35 @@
+final class SpinnerTableCell: StyledTableCell {
+
+    let activity = UIActivityIndicatorView(activityIndicatorStyle: .gray)
+    var indicatorHolder: UITableViewCellAccessoryType?
```
Are you using this outside this class? If not, make it private :)

jdisho commented on Classes/Views/SpinnerTableCell.swift:
```diff
@@ -0,0 +1,35 @@
+        activity.hidesWhenStopped = true
+    }
+
+    public func showSpinner() {
```
Why `public` and not `internal` (default) ?

jdisho commented on Classes/Views/SpinnerTableCell.swift:
```diff
@@ -0,0 +1,35 @@
+        self.accessoryType = .none
+        activity.startAnimating()
+    }
+    public func stopSpinner() {
```
Why `public` and not `internal` (default) ?

Sherlouk commented on Classes/New Issue/NewIssueTableViewController.swift:
```diff
@@ -69,17 +70,22 @@ final class NewIssueTableViewController: UITableViewController, UITextFieldDeleg
+        signature: IssueSignatureType? = nil
+        ) -> NewIssueTableViewController? {
+        
+        let storyboard = UIStoryboard(name: NSLocalizedString("NewIssue", comment: ""), bundle: nil)
```
Storyboard name definitely shouldn't be localised!

Sherlouk commented on Classes/Views/Constants.swift:
```diff
@@ -52,5 +71,7 @@ enum Constants {
         static let reviewers = NSLocalizedString("Reviewers", comment: "")
         static let reviewGitHubAccess = NSLocalizedString("Review GitHub Access", comment: "")
         static let clear = NSLocalizedString("Clear", comment: "")
+        static let emptyString = NSLocalizedString("", comment: "")
```
Empty string constant shouldn't be localized either

Sherlouk commented on Classes/Utility/String+Regex.swift:
```diff
@@ -0,0 +1,32 @@
+                )
+            )
+            return results.map { str.substring(with: $0.range) }
+        } catch let error as NSError {
```
Unnecessary cast to NSError here

Sherlouk commented on Classes/Repository/RepositoryViewController.swift:
```diff
@@ -147,21 +147,17 @@ ContextMenuDelegate {
-
-        newIssueViewController.delegate = self
-        weak var weakSelf = self
+        let action = UIAlertAction(title: "New Issue", style: .default) { _ in
```
Title should be localized

Sherlouk commented on Classes/New Issue/GithubClient+Template.swift:
```diff
@@ -0,0 +1,130 @@
+            owner: repo.owner,
+            repo: repo.name,
+            branch: repo.defaultBranch,
+            path: "\(Constants.Filepaths.githubIssue)/\(filename)") { (result) in
```
If this gets rebased onto `master` I believe @rnystrom has created a URLBuilder to make this a bit nicer

Huddie replied:
I think URLBuilder makes a full url (Https/http:// ... ). I'll check it out

Huddie replied:
Ya URLBuilder seems to require a scheme. I also don't see why URLBuilder is better than a constant. Mind if I keep this for now?

Sherlouk replied:
Not at all, hadn't looked into the URLBuilder stuff and wondered if it would've worked! All good for now

Sherlouk commented on Classes/New Issue/IssueTemplateHelper.swift:
```diff
@@ -0,0 +1,114 @@
+        let alertView = UIAlertController(
+            title: NSLocalizedString("New Issue", comment: ""),
+            message: NSLocalizedString("Choose Template", comment: ""),
+            preferredStyle: .alert
```
I'm thinking action sheet might be better here, but this is personal preference

Huddie replied:
Bas also mentioned this, I'll switch for now and we can switch back if need be

Sherlouk commented on Classes/Settings/Settings.storyboard:
```diff
@@ -125,7 +125,7 @@
                             <tableViewSection headerTitle="🦅 Githawk" id="3QH-oa-l2C">
                                 <cells>
-                                    <tableViewCell clipsSubviews="YES" contentMode="scaleToFill" selectionStyle="default" accessoryType="disclosureIndicator" indentationWidth="10" textLabel="kdx-s1-NXl" style="IBUITableViewCellStyleDefault" id="WwO-m1-s3q" customClass="StyledTableCell" customModule="Freetime" customModuleProvider="target">
+                                    <tableViewCell clipsSubviews="YES" contentMode="scaleToFill" selectionStyle="default" accessoryType="disclosureIndicator" indentationWidth="10" textLabel="kdx-s1-NXl" style="IBUITableViewCellStyleDefault" id="WwO-m1-s3q" customClass="SpinnerTableCell" customModule="Freetime" customModuleProvider="target">
```
Title here hasn't been changed, did we want to do that?

Huddie replied:
I subclassed StyledTableCell to allow an activity indicator.

Sherlouk replied:
I was more referring to the label inside of the cell, but can't add a review to a line outside of the change context 😉

Huddie replied:
Oh haha! Ya ill fix that 😄Sorry

Sherlouk commented:
This looks good to me! 🎉 Great work!

Huddie commented:
@Sherlouk Thanks for the help!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

jdisho commented:
Great work @Huddie 🎉

Huddie commented:
@rnystrom  I'm moving strings out of constants that aren't used elsewhere but I kind of like having the different regex stored in one place so we can see them easily. I know we've updated a few regex already and having them in one place could be nice. Should I:

- Move all regex out of constants back into the code
- Keep in the constants file
- Move to a new file where all regex is held (enum Regex .. )

?

rnystrom commented:
Hm. If the regret isn’t used again I say keep it inline. IMO its _more_ readable when it’s lives where it’s used.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@rnystrom Once I updated constants to remove
the unnecessary stuff I mentioned above is this good to go?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
This is the updated gif. 

<img src=https://user-images.githubusercontent.com/10662471/48520881-4a7c2100-e840-11e8-8b69-e2cc5cc01799.gif width=250>

BrianLitwin commented:
@Huddie Sorry I just saw your comment above- 
Tbh, I rely on unit tests with the regex's, not a deep understanding of regular expressions or Swift's implementation. 
These files all follow a similar pattern: 

[String+GithubEmoji](https://github.com/GitHawkApp/GitHawk/blob/master/Classes/Issues/Comments/Markdown/String+GitHubEmoji.swift)

[String+Shortlink](https://github.com/GitHawkApp/GitHawk/blob/master/Classes/Issues/Comments/Markdown/String+Shortlink.swift)

[String+StripHTMLComments](https://github.com/GitHawkApp/GitHawk/blob/master/Classes/Issues/Comments/Markdown/String+StripHTMLComments.swift)

[String + CustomBuilding](https://github.com/GitHawkApp/GitHawk/blob/master/Classes/Issues/Comments/Markdown/String+CustomBuilding.swift)

You'll see the pattern quickly: a public String method that returns what you need, eg `var templateNameAndDescription`, plus a single file with private properties and functions that contain all the logic. 

```Swift 

private let nameRegex = Try! NSRegularExpression(pattern: )
private let aboutRegex = Try! NSRegularExpression(pattern: )

extension String {
        var templateNameAndDescription: (name: String, description: String)? {
               // logic 
         }
}
```

Huddie commented:
Cool. I’m going to whip up a few unit tests and check out those files 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
Does this need updating?

rnystrom commented:
I haven't been following along b/c of the other eyeballs that have been on this. Lmk if/when I should give it a look.

Huddie commented:
@rnystrom if you don’t mind taking a look through I’d appreciate it. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@rnystrom Anything fixes required?

rnystrom commented:
Haven’t gotten to a review yet 😞

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

Huddie commented:
@rnystrom all good. Had to fix merge problems anyways 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

rnystrom commented on FreetimeTests/TemplateRegexTests.swift:
```diff
@@ -0,0 +1,123 @@
+        """
+
+        var cleanedFile = ""
+        if let textToClean = template.matches(regex: "([-]{3,})([\\s\\S]*)([-]{3,})").first {
```
Regex should be shared w/ actual implementation!

Huddie commented on Classes/New Issue/GithubClient+Template.swift:
```diff
@@ -0,0 +1,150 @@
+    let repoDetails: IssueTemplateRepoDetails
+    let session: GitHubUserSession?
+    let viewController: UIViewController
+    weak var delegate: NewIssueTableViewControllerDelegate?
```
@rnystrom We can't reduce the networking calls because the title and description of files are at the top of the files. The file names are not equal to the template names. Let me know if that makes sense. I will make the changes regarding passing the templates to the VC and letting it take care of displaying the alert.

Huddie commented:
@rnystrom I'll take care of the nits and feedback. I replied above in regads to the number of networking calls. I don't think it can be reduced unfortuantly. 
If no templates are found we just open NewIssueViewController with nothing passed in. Basically a blank issue.

Huddie commented on Classes/New Issue/IssueTemplateHelper.swift:
```diff
@@ -0,0 +1,108 @@
+                        viewController.delegate = details.delegate
+                        let navController = UINavigationController(rootViewController: viewController)
+                        navController.modalPresentationStyle = .formSheet
+                        details.viewController.route_present(to: navController)
```
@rnystrom Can you explain this a bit? If I switch to `route_present(to: viewController)` no navigation bar shows in the pushed view. I think I may not fully understand.

Huddie replied:
Also separate question: Should it be route_detail? on iPad it pushes the new VC over the entire screen not into the detail section

Huddie commented:
@rnystrom Ready for another review whenever. Thanks for the last one!

Huddie commented:
@rnystrom 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

rnystrom commented on Classes/New Issue/GithubClient+Template.swift:
```diff
@@ -0,0 +1,115 @@
+import Squawk
+
+private let githubIssueURL = ".github/ISSUE_TEMPLATE"
+let tempalateRegex = ".github/ISSUE_TEMPLATE"
```
`private`?

Sherlouk replied:
also nit on the spelling 😄

rnystrom commented on Classes/New Issue/GithubClient+Template.swift:
```diff
@@ -0,0 +1,115 @@
+        ) {
+
+        // For Testing.
+        if let testingFile = testingFile { completion(.success(testingFile)) }
```
Let's newline the closure execution here to stay consistent w/ other `if`s

rnystrom commented on Classes/New Issue/GithubClient+Template.swift:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+
+    func createTemplate(
```
nit: this needs a new name, I keep thinking that it's actually creating something. Can it just use the "fetch" naming used elsewhere?

rnystrom commented on Classes/New Issue/GithubClient+Template.swift:
```diff
@@ -0,0 +1,115 @@
+                case .success(let files):
+                    for file in files {
+                        templateGroup.enter()
+                        self.createTemplate(repoDetails: repoDetails, filename: file.name, completion: { result in
```
I'm going to block this PR on the multiple-requests part. It's just too wasteful IMO as-is. Some options:

1. Do a fetch for the directory and just show the file names in the list instead of the template title. Then fetch the contents of the template once its been selected.
2. Only fetch this info once then cache it. Always use the cache but allow the user to pull-to-refresh the templates list.

I favor option 1 personally.

Sherlouk replied:
Sorry to jump in on here @rnystrom, and I know it's a massive shift in direction but could I suggest a third option here?

If we use the v4 API (ooh bois he's said the danger word) we can actually do this with a single request which returns the contents of every template within the directory?

This request: (Notably, we would need to replace the default branch and repo owner/name)
```gql
query {
  repository(owner: "GitHawkApp", name:"GitHawk") {
    object(expression: "master:.github/ISSUE_TEMPLATE") {
      ... on Tree {
        entries {
          name
          object {
            ... on Blob {
              text
            }
          }
        }
      }
    }
  }
}
```

Returns us this response:
```json
{
  "data": {
    "repository": {
      "object": {
        "entries": [
          {
            "name": "bug_report.md",
            "object": {
              "text": "---\nname: Bug report  (etc etc, I've truncated this)"
            }
          },
          {
            "name": "feature_request.md",
            "object": {
              "text": "---\nname: Feature request (etc etc, I've truncated this)"
            }
          }
        ]
      }
    }
  }
}
```

One request, all the information? Opinion Ryan?

rnystrom replied:
Lovd it!
